From d8526b8d60470e34be9bea0ace215ca1dfbb8f89 Mon Sep 17 00:00:00 2001
From: Pantelis Antoniou <panto@antoniou-consulting.com>
Date: Wed, 30 Jan 2013 20:09:11 +0200
Subject: [PATCH 097/177] tilcdc: Add reduced blanking mode checks

The driver doesn't support (yet?) non-reduced blanking mode.
So by default disable any non-reduced blanking modes, and add a switch
to enable them in the future.

Signed-off-by: Pantelis Antoniou <panto@antoniou-consulting.com>
---
 drivers/gpu/drm/tilcdc/tilcdc_crtc.c | 11 +++++++++++
 drivers/gpu/drm/tilcdc/tilcdc_drv.c  |  5 ++++-
 drivers/gpu/drm/tilcdc/tilcdc_drv.h  |  1 +
 3 files changed, 16 insertions(+), 1 deletion(-)

diff --git a/drivers/gpu/drm/tilcdc/tilcdc_crtc.c b/drivers/gpu/drm/tilcdc/tilcdc_crtc.c
index 5dd3c7d..628d9aa 100644
--- a/drivers/gpu/drm/tilcdc/tilcdc_crtc.c
+++ b/drivers/gpu/drm/tilcdc/tilcdc_crtc.c
@@ -422,6 +422,7 @@ int tilcdc_crtc_mode_valid(struct drm_crtc *crtc, struct drm_display_mode *mode)
 {
 	struct tilcdc_drm_private *priv = crtc->dev->dev_private;
 	unsigned int bandwidth;
+	int rb;
 
 	if (mode->hdisplay > tilcdc_crtc_max_width(crtc))
 		return MODE_VIRTUAL_X;
@@ -438,6 +439,16 @@ int tilcdc_crtc_mode_valid(struct drm_crtc *crtc, struct drm_display_mode *mode)
 	if (bandwidth > priv->max_bandwidth)
 		return MODE_BAD;
 
+	if (priv->allow_non_rblank == 0) {
+		/* we only support reduced blanking modes */
+		rb = (mode->htotal - mode->hdisplay == 160) &&
+		       (mode->hsync_end - mode->hdisplay == 80) &&
+		       (mode->hsync_end - mode->hsync_start == 32) &&
+		       (mode->vsync_start - mode->vdisplay == 3);
+		if (!rb)
+			return MODE_BAD;
+	}
+
 	return MODE_OK;
 }
 
diff --git a/drivers/gpu/drm/tilcdc/tilcdc_drv.c b/drivers/gpu/drm/tilcdc/tilcdc_drv.c
index 3ffa2e5..0fe14ea 100644
--- a/drivers/gpu/drm/tilcdc/tilcdc_drv.c
+++ b/drivers/gpu/drm/tilcdc/tilcdc_drv.c
@@ -235,9 +235,12 @@ static int tilcdc_load(struct drm_device *dev, unsigned long flags)
 	}
 #endif
 
-	if (of_property_read_u32(node, "max-bandwidth", &priv->max_bandwidth))
+	if (of_property_read_u32(node, "ti,max-bandwidth", &priv->max_bandwidth))
 		priv->max_bandwidth = 1280 * 1024 * 60;
 
+	priv->allow_non_rblank = of_property_read_bool(node,
+			"ti,allow-non-reduced-blanking-modes");
+
 	pm_runtime_enable(dev->dev);
 
 	/* Determine LCD IP Version */
diff --git a/drivers/gpu/drm/tilcdc/tilcdc_drv.h b/drivers/gpu/drm/tilcdc/tilcdc_drv.h
index 8242b5a..c858f73 100644
--- a/drivers/gpu/drm/tilcdc/tilcdc_drv.h
+++ b/drivers/gpu/drm/tilcdc/tilcdc_drv.h
@@ -43,6 +43,7 @@ struct tilcdc_drm_private {
 
 	/* don't attempt resolutions w/ higher W * H * Hz: */
 	uint32_t max_bandwidth;
+	int allow_non_rblank;	/* ATM we don't support non reduced blank modes */
 
 	/* register contents saved across suspend/resume: */
 	u32 saved_register[12];
-- 
1.8.1.4

